src/frontegg-auth/src/client.rs

```diff
@@ -44,6 +59,167 @@ impl Client {
             .with(RetryTransientMiddleware::new_with_policy(retry_policy))
             .build();
 
-        Self { client }
+        let inflight_requests = Arc::new(Mutex::new(HashMap::new()));
+
+        Self {
+            client,
+            inflight_requests,
+        }
+    }
+
+    /// Makes a request to the provided URL, possibly de-duping by attaching a listener to an
+    /// already in-flight request.
+    async fn make_request<Req, Resp>(&self, url: String, req: Req) -> Result<Resp, Error>
+    where
+        Req: AuthRequest,
+        Resp: AuthResponse,
+    {
+        let req = req.into_request();
+
+        // Note: we get the reciever in a block to scope the access to the mutex.
+        let rx = {
+            let mut inflight_requests = self
+                .inflight_requests
+                .lock()
+                .expect("Frontegg Auth Client panicked");
+            let (tx, rx) = tokio::sync::oneshot::channel();
+
+            match inflight_requests.get_mut(&req) {
+                // Already have an inflight request, add to our list of waiters.
+                Some(senders) => {
+                    tracing::debug!("reusing request, {req:?}");
+                    senders.push(tx);
+                    rx
+                }
+                // New request! Need to queue one up.
+                None => {
+                    tracing::debug!("spawning new request, {req:?}");
+
+                    inflight_requests.insert(req.clone(), vec![tx]);
+
+                    let client = self.client.clone();
+                    let inflight = Arc::clone(&self.inflight_requests);
+                    let req_ = req.clone();
+
+                    mz_ore::task::spawn(move || "frontegg-auth-request", async move {
+                        // Make the actual request.
+                        let result = async {
+                            let resp = client
+                                .post(&url)
+                                .json(&req_.into_json())
+                                .send()
+                                .await?
+                                .error_for_status()?
+                                .json::<Resp>()
+                                .await?;
+                            Ok::<_, Error>(resp)
+                        }
+                        .await;
+
+                        // Get all of our waiters.
+                        let mut inflight = inflight.lock().expect("Frontegg Auth Client panicked");
+                        let Some(waiters) = inflight.remove(&req) else {
+                            tracing::error!("Inflight entry already removed? {req:?}");
+                            return;
+                        };
+
+                        // Tell all of our waiters about the result.
+                        let response = result.map(|r| r.into_response());
+                        for tx in waiters {
+                            let _ = tx.send(response.clone());
```

Comment:
Thoughts on Arc'ing the response to avoid so many clones?

Reply:
Yeah I thought about wrapping the response and error in `Arc`s, but that would require us to return `Arc<Response>` to callers, which I didn't think was worth it because the callers were all using owned versions of the structs, so we'd need to clone them anyways. Happy to change this if you want me to